### PR TITLE
Add verification and public key importing 

### PR DESCRIPTION
**Pull request by AndrzejKurek**
This PR is a prerequisite for https://github.com/ARMmbed/mbed-os-example-atecc608a/pull/3.

### CODE REVIEW

File: atecc608a_se.c [hunk 1]
```diff
@@ -60,7 +60,7 @@
 
 static ATCAIfaceCfg atca_iface_config = {
     .iface_type = ATCA_I2C_IFACE,
-    .devtype = ATECC608A,
+    .devtype = ATECC508A,
```
Comment (Patater) [inline atecc608a_se.c:65]: Why was this change necessary? Explain the motivation in the commit message body. It's unclear if there was anything broken or misused by having this be 608A.

File: atecc608a_se.c [hunk 2]
```diff
@@ -161,6 +161,43 @@ static psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
...
+                                                psa_algorithm_t alg,
+                                                psa_key_usage_t usage,
+                                                const uint8_t *p_data,
+                                                size_t data_length)
+{
+    const uint16_t key_id = key_slot;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    /* Keys 8 to 15 can store public keys. Slots 1-7 are too small. */
+    if (key_id < 7 || key_id > 15 || data_length != ATCA_PUB_KEY_SIZE)
```
Comment (Patater) [inline atecc608a_se.c]: This isn't device configuration specific, right? This is an inherent property of the device itself, no matter how the device is configured.

File: atecc608a_se.c [hunk 3]
```diff
@@ -161,6 +161,43 @@ static psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
...
+    if (alg != PSA_ALG_ECDSA(PSA_ALG_SHA_256) && alg != PSA_ALG_ECDSA_ANY)
+    {
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    ASSERT_SUCCESS_PSA(atecc608a_init());
+
+    /* Signature format is R and S integers in big-endian format.
+     * 64 bytes for P256 curve. */
+    ASSERT_SUCCESS(atcab_write_pubkey(key_id, p_data));
```
Comment (Patater) [inline atecc608a_se.c]: Do we need to strip off the `04` in the front of `p_data` before passing on to the driver?

File: atecc608a_se.c [hunk 4]
```diff
@@ -208,20 +245,64 @@ static psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
...
+                                         size_t hash_length,
+                                         const uint8_t *p_signature,
+                                         size_t signature_length)
+{
+    const uint16_t key_id = key_slot;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+    bool is_verified = false;
+
+    /* Keys 8 to 15 can store public keys. */
+    if (key_id < 7 || key_id > 15)
```
Comment (Patater) [inline atecc608a_se.c]: Again, this is not particular to our configuration, is it?

File: atecc608a_se.c [hunk 5]
```diff
@@ -208,20 +245,64 @@ static psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
...
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+
+    /* We can only do ECDSA on SHA-256 */
+    if (alg != PSA_ALG_ECDSA(PSA_ALG_SHA_256) && alg != PSA_ALG_ECDSA_ANY)
+    {
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    if (hash_length != 32)
```
Comment (Patater) [inline atecc608a_se.c]: Should we use the hash length macro here? `PSA_HASH_SIZE(PSA_ALG_SHA_256)`

May want to update the signature function as well at the same time, in a separate commit.

File: atecc608a_se.c [hunk 6]
```diff
@@ -208,20 +245,64 @@ static psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
...
+    return status;
+}
 
 #define PSA_ATECC608A_LIFETIME 0xdeadbeefU
 
 static psa_drv_se_asymmetric_t atecc608a_asymmetric =
 {
     .p_sign = &atecc608a_asymmetric_sign,
-    .p_verify = 0,
+    .p_verify = &atecc608a_asymmetric_verify,
```
Comment (Patater) [inline atecc608a_se.c]: The `&` here isn't necessary. I'd recommend removing it. I would also recommend removing it for `atecc608a_asymmetric_sign` in a separate commit.

File: atecc608a_se.c [hunk 7]
```diff
@@ -208,20 +245,64 @@ static psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
...
     .p_encrypt = 0,
     .p_decrypt = 0,
 };
 
 static psa_drv_se_key_management_t atecc608a_key_management =
 {
-    .p_import = 0,
+    /* So far there is no public key import function in the API, so use this instead */
+    .p_import = &atecc608a_import_public_key,
```
Comment (Patater) [inline atecc608a_se.c]: No `&` please.

**Review by Patater — CHANGES_REQUESTED** [review]

File: atecc608a_se.c (hunk 2 above)
Comment (AndrzejKurek) [reply atecc608a_se.c]: Yes - it's about the slot sizes - slots 0-7 have 32 bytes of size, 9-15 - 72, slot 8 - 416.

File: atecc608a_se.c (hunk 3 above)
Comment (AndrzejKurek) [reply atecc608a_se.c]: We can do it here, you're right.

File: atecc608a_se.c (hunk 5 above)
Comment (AndrzejKurek) [reply atecc608a_se.c]: Putting `PSA_HASH_SIZE` here may suggest that it has something to do with PSA. I would use a `ATCA_`-something macro, but I'm not sure if there's any... maybe `ATCA_SHA_DIGEST_SIZE`?

[The cryptoauthlib code](https://github.com/MicrochipTech/cryptoauthlib/blob/76ccc87aebf3ac10dca99317fed57ee1921b2ef5/lib/basic/atca_basic_verify.c#L371) uses 32 as a hardcoded value, with no prior definition.

Comment (AndrzejKurek) [reply atecc608a_se.c]: I rethought it overnight, and I guess that `PSA_HASH_SIZE(PSA_ALG_SHA_256)` is better than a raw 32 - people who are more proficient than me will catch it immediately.

**Comment by AndrzejKurek** [comment]
@Patater - I've addressed your comments. Please re-review.

File: atecc608a_se.c [hunk 8]
```diff
@@ -172,8 +172,9 @@ static psa_status_t atecc608a_import_public_key(psa_key_slot_number_t key_slot,
     const uint16_t key_id = key_slot;
     psa_status_t status = PSA_ERROR_GENERIC_ERROR;
 
-    /* Keys 8 to 15 can store public keys. Slots 1-7 are too small. */
-    if (key_id < 7 || key_id > 15 || data_length != ATCA_PUB_KEY_SIZE)
+    /* Keys 8 to 15 can store public keys. Slots 1-7 are too small.
+       We also check if the key has a size of 65 {0x04, X, Y}. */
```
Comment (Patater) [inline atecc608a_se.c]: Style: Add a `*` before `We`, aligned with the above `*`.

File: atecc608a_se.c [hunk 9]
```diff
@@ -161,6 +163,43 @@ static psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
...
+                                                psa_key_usage_t usage,
+                                                const uint8_t *p_data,
+                                                size_t data_length)
+{
+    const uint16_t key_id = key_slot;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    /* Keys 8 to 15 can store public keys. Slots 1-7 are too small.
+       We also check if the key has a size of 65 {0x04, X, Y}. */
+    if (key_id < 7 || key_id > 15 || data_length != ATCA_PUB_KEY_SIZE + 1)
```
Comment (Patater) [inline atecc608a_se.c]: What we care to check is that the input to this function is in PSA format for the only algorithm we support. We should use `PSA_ASYMMETRIC_SIGN_OUTPUT_SIZE()` instead of `ATCA_PUB_KEY_SIZE+1`, as the PSA format may change. We want to always use the PSA format as input to this function (so it makes sense to use its size macros when checking input), and always use `ATCA_PUB_KEY_SIZE` as output to the SE (when checking that after we've transformed the input to the SE's format, it's the correct size for the SE.)

**Review by Patater — CHANGES_REQUESTED** [review]

File: atecc608a_se.c (hunk 9 above)
Comment (AndrzejKurek) [reply atecc608a_se.c]: But you're speaking about the size of a buffer for a signature using a given key (and for our case it's `PSA_ASYMMETRIC_SIGN_OUTPUT_SIZE(PSA_KEY_TYPE_ECC_PUBLIC_KEY(PSA_ECC_CURVE_SECP256R1), 256, PSA_ALG_ECDSA(PSA_ALG_SHA_256))` == 64), not the size of a public key exported from PSA, which is 65. Maybe you've meant `PSA_KEY_EXPORT_ECC_PUBLIC_KEY_MAX_SIZE`?

Comment (Patater) [reply atecc608a_se.c]: Yes, sorry. Wrong macro. Does `PSA_KEY_EXPORT_ECC_PUBLIC_KEY_MAX_SIZE` take an algorithm parameter? Is it overly large?

Comment (AndrzejKurek) [reply atecc608a_se.c]: It's short - added.

File: atecc608a_se.c [hunk 10]
```diff
@@ -161,6 +163,43 @@ static psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
...
+                                                psa_key_usage_t usage,
+                                                const uint8_t *p_data,
+                                                size_t data_length)
+{
+    const uint16_t key_id = key_slot;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    /* Keys 8 to 15 can store public keys. Slots 1-7 are too small.
+     * We also check if the key has a size of 65 {0x04, X, Y}. */
+    if (key_id < 7 || key_id > 15 || data_length != PSA_KEY_EXPORT_ECC_PUBLIC_KEY_MAX_SIZE(256))
```
Comment (gilles-peskine-arm) [inline atecc608a_se.c]: `PSA_KEY_EXPORT_ECC_PUBLIC_KEY_MAX_SIZE` is an internal macro. (You can tell because it doesn't have Doxygen documentation… Yes, I know, this should be indicated more clearly.) Please stick to the standard macro `PSA_KEY_EXPORT_MAX_SIZE`.

By the way, the check on the key size is redundant with checking the key type: the core is guaranteed to pass a valid key here. You can leave it in for robustness, or omit it to save a couple of bytes of code size.

File: atecc608a_se.c [hunk 11]
```diff
@@ -161,6 +163,43 @@ static psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
...
+    {
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (!PSA_KEY_TYPE_IS_PUBLIC_KEY(type))
```
Comment (gilles-peskine-arm) [inline atecc608a_se.c]: This function doesn't accept any public key: it needs a P256 key. Test `type == PSA_KEY_TYPE_ECC_PUBLIC_KEY(PSA_ECC_CURVE_SECP256R1)`. Also the error here should be `NOT_SUPPORTED` rather than `INVALID_ARGUMENT`.

File: atecc608a_se.c (hunk 10 above)
Comment (gilles-peskine-arm) [inline atecc608a_se.c]: Please put they `key_id` checks and the `data_length` check in separate `if` statements. Combining unrelated checks makes the code harder to read, causes coverage measurements to give less useful information, and makes the reason for an error harder to trace in a debugger.

File: atecc608a_se.c [hunk 12]
```diff
@@ -161,6 +163,43 @@ static psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
...
+    {
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+
+    /* We can only do ECDSA on SHA-256 */
```
Comment (gilles-peskine-arm) [inline atecc608a_se.c]:
```suggestion
    /* We can only do randomized ECDSA on SHA-256 */
```

File: atecc608a_se.c [hunk 13]
```diff
@@ -179,7 +218,7 @@ static psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
         return PSA_ERROR_NOT_SUPPORTED;
     }
 
-    if (hash_length != 32)
+    if (hash_length != PSA_HASH_SIZE(PSA_ALG_SHA_256))
     {
         /* The driver only supports signing things of length 32. */
```
Comment (gilles-peskine-arm) [inline atecc608a_se.c:233]: The comment says “up to” but the code says “exactly”.

File: atecc608a_se.c [hunk 14]
```diff
@@ -208,24 +247,68 @@ static psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
...
+                                         const uint8_t *p_hash,
+                                         size_t hash_length,
+                                         const uint8_t *p_signature,
+                                         size_t signature_length)
+{
+    const uint16_t key_id = key_slot;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+    bool is_verified = false;
+
+    /* Keys 8 to 15 can store public keys. */
```
Comment (gilles-peskine-arm) [inline atecc608a_se.c]: “can store” — but why does this matter at this point? What matters here is what _is_ stored. Please clarify the comment or improve the code. Is this a sanity check or does it prevent failures later?

File: atecc608a_se.c [hunk 15]
```diff
@@ -208,24 +247,68 @@ static psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
...
+    if (hash_length != PSA_HASH_SIZE(PSA_ALG_SHA_256))
+    {
+        /* The driver only supports hashes of length 32. */
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    if (signature_length != ATCA_SIG_SIZE)
+    {
+        /* The driver only supports signatures of length 64. */
+        return PSA_ERROR_NOT_SUPPORTED;
```
Comment (gilles-peskine-arm) [inline atecc608a_se.c]: A signature whose length isn't 64 would be invalid in any implementation, so `INVALID_SIGNATURE` would be the right error code here.

File: atecc608a_se.c [hunk 16]
```diff
@@ -208,24 +247,68 @@ static psa_status_t atecc608a_asymmetric_sign(psa_key_slot_number_t key_slot,
...
+    bool is_verified = false;
+
+    /* Keys 8 to 15 can store public keys. */
+    if (key_id < 7 || key_id > 15)
+    {
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+
+    /* We can only do ECDSA on SHA-256 */
+    if (alg != PSA_ALG_ECDSA(PSA_ALG_SHA_256) && alg != PSA_ALG_ECDSA_ANY)
```
Comment (gilles-peskine-arm) [inline atecc608a_se.c:274]: This check is redundant with the similar check at import time: the core is guaranteed to pass an algorithm that complies with the policy. You can leave it in for robustness, or omit it to save a couple of bytes of code size.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]

File: atecc608a_se.c (hunk 14 above)
Comment (AndrzejKurek) [reply atecc608a_se.c]: This is a sanity check for arguments - we haven't added any validation of what exactly _is_ pre-provisioned on the chip. Should we?

Comment (gilles-peskine-arm) [reply atecc608a_se.c]: It depends primarily on what the consequences of an error are, and secondarily on how costly it is to make robust checks. At this point, how pre-provisioning works is still in an exploratory stage, so I don't have any particular requirements on how it's validated.

Here are some questions that are relevant for deciding what to do. Answering them is not a blocker for this PR, only a blocker for putting the driver into production. What else might be in slots 8–15 if people follow the driver documentation? What configuration indicates what is present in these slots and who owns these slots? Who can modify this configuration? What happens if the content of the slot does not have the right type (crash, nonsensical output, nonsensical output that leaks confidential data, …)?

File: atecc608a_se.c (hunk 13 above)
Comment (AndrzejKurek) [reply atecc608a_se.c:233]: Which comment exactly do you mean?

Comment (gilles-peskine-arm) [reply atecc608a_se.c:233]: I don't know. I must have been looking at some old commit.

File: atecc608a_se.c (hunk 14 above)
Comment (AndrzejKurek) [reply atecc608a_se.c]: There's no driver documentation on provisioning so far. 
Initially, the slots are only limited by their size and the fact that slot 15 has an unique use counter:
- Slots 0-7 have 36 bytes (usually used for private keys);
- Slot 8 has 416 bytes (usually used for a certificate / signature / arbitrary data);
- Slots 9-14 have 72 bytes (usually used for public keys);
- Slot 15 has 72 bytes and can be a subject of a limited use counter.

Key slots may be used by the hardware only if the configuration is locked. Configuration can be read later on to determine how are the slots configured.
If an illegal operation is performed using a slot, or a slot contains bad data, an error is returned from the API.
Edit: that's also a thing to consider in the SE driver API - maybe slot configuration should be done on a struct-per-slot basis?

Comment (gilles-peskine-arm) [reply atecc608a_se.c]: Thanks. So the check should be for slots 9–14 here. Also, please make this check an auxiliary function, since it's used in two places already.

Comment (AndrzejKurek) [reply atecc608a_se.c]: As this is a sanity, not a configuration check, I'd rather opt for leaving the 8-15 check here - someone can have the board configured so that the public key is in slot 8 or 15. On the contrary, it's not possible to have a key in slot 0-7.

Comment (AndrzejKurek) [reply atecc608a_se.c]: I've added a function, and there was a typo in the numbers anyway - fixed.

**Review by gilles-peskine-arm — DISMISSED** [review]

File: atecc608a_se.c (hunk 16 above)
Comment (Patater) [reply atecc608a_se.c:274]: How will the core know what policy the driver/hardware is physically capable of implementing? Shouldn't the driver be the source of this information and be able to reject it (either at policy setting time [for which we'd need new driver function pointers] or at use time [here])?

File: atecc608a_se.c [hunk 17]
```diff
@@ -161,6 +169,45 @@ static psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
...
+
+    /* We can only do randomized ECDSA on SHA-256 */
+    if (alg != PSA_ALG_ECDSA(PSA_ALG_SHA_256) && alg != PSA_ALG_ECDSA_ANY)
+    {
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    ASSERT_SUCCESS_PSA(atecc608a_init());
+
+    /* PSA public key format is {0x04, X, Y}, and we want just raw {X,Y}. */
```
Comment (Patater) [inline atecc608a_se.c]: "`we want just raw {X,Y}`" is ambiguous. Suggest to replace "we want" with "the hardware wants" or "the hardware works with".

File: atecc608a_se.c [hunk 18]
```diff
@@ -161,6 +169,45 @@ static psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
...
+    ASSERT_SUCCESS(atcab_write_pubkey(key_id, p_data+1));
```
Comment (Patater) [inline atecc608a_se.c]: Style: Add spaces around binary operators like `+`.

File: mbed-cryptoauthlib.lib [hunk 19]
```diff
@@ -1 +1 @@
-https://github.com/janjongboom/mbed-cryptoauthlib/#c74828b1bbdf67281ad4689b05d88b2b32e86c75
+https://github.com/ARMmbed/mbed-cryptoauthlib/#6c0cacc6bb1428cff8b7e252c86ecc0f03d83ce7
```
Comment (Patater) [inline mbed-cryptoauthlib.lib]: Style: Should there be a newline at the end of this file? It looks like there used to be one, but we've removed it.

File: atecc608a_se.c (hunk 16 above)
Comment (gilles-peskine-arm) [reply atecc608a_se.c:274]: The driver interface needs to change to allow the driver code to pick a slot number when creating a key. We have some architecture discussions on the topic but no concrete header files yet.

File: atecc608a_se.c [hunk 20]
```diff
@@ -161,6 +169,45 @@ static psa_status_t atecc608a_export_public_key(psa_key_slot_number_t key,
...
+                                                const uint8_t *p_data,
+                                                size_t data_length)
+{
+    const uint16_t key_id = key_slot;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    ASSERT_SUCCESS_PSA(is_public_key_slot(key_id));
+
+    /* Check if the key has a size of 65 {0x04, X, Y}. */
+    if(data_length != PSA_KEY_EXPORT_MAX_SIZE(PSA_KEY_TYPE_ECC_PUBLIC_KEY(PSA_ECC_CURVE_SECP256R1),
```
Comment (Patater) [inline atecc608a_se.c]: Style: Add space after `if` before `(`.

Style: this line may be over 79 characters long.

**Review by Patater — CHANGES_REQUESTED** [review]

**Review by gilles-peskine-arm — APPROVED** [review]

**Review by Patater — APPROVED** [review]
